torch/_lobpcg.py

```diff
@@ -13,6 +13,336 @@
+        )
+    ).matmul(Ut)
+
+    return res
+
+def _symeig_backward(D_grad, U_grad, A, D, U, largest):
+    # if `U` is square, then the columns of `U` is a complete eigenspace
+    if U.size(-1) == U.size(-2):
+        return _symeig_backward_complete_eigenspace(
+            D_grad, U_grad, A, D, U
+        )
+    else:
+        return _symeig_backward_partial_eigenspace(
+            D_grad, U_grad, A, D, U, largest
+        )
+
+class LOBPCGAutogradFunction(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx,
+                A,              # type: Tensor
+                k,              # type: Optional[int]
+                B,              # type: Optional[Tensor]
+                X,              # type: Optional[Tensor]
+                n,              # type: Optional[int]
+                iK,             # type: Optional[Tensor]
+                niter,          # type: Optional[int]
+                tol,            # type: Optional[float]
+                largest,        # type: Optional[bool]
+                method,         # type: Optional[str]
+                tracker,        # type: Optional[None]
+                ortho_iparams,  # type: Optional[Dict[str, int]]
+                ortho_fparams,  # type: Optional[Dict[str, float]]
+                ortho_bparams,  # type: Optional[Dict[str, bool]]
+                ):
+        # type: (...) -> Tuple[Tensor, Tensor]
+
+        # makes sure that input is contiguous for efficiency.
+        # Note: autograd does not support dense gradients for sparse input yet.
+        A = A.contiguous() if (not A.is_sparse) else A
+        if B is not None:
+            B = B.contiguous() if (not B.is_sparse) else B
+
+        D, U = _lobpcg(
+            A, k, B, X,
+            n, iK, niter, tol, largest, method, tracker,
+            ortho_iparams, ortho_fparams, ortho_bparams
+        )
+        # D, U = torch.symeig(A, eigenvectors=True, upper=True)
+        # D = D[..., -k:]
+        # U = U[..., :, -k:]
+
+        D = D.contiguous()
+        U = U.contiguous()
+        ctx.save_for_backward(A, B, D, U, largest)
+
+        return D, U
+
+    @staticmethod
+    def backward(ctx, D_grad, U_grad):
+        A_grad = B_grad = None
+        grads = [None] * 14
+
+        A, B, D, U, largest = ctx.saved_tensors
+
+        # lobpcg.backward has some limitations. Checks for insupported input
+        if A.is_sparse or (B is not None and B.is_sparse and B.requires_grad):
+            raise ValueError(
+                'lobpcg.backward does not support sparse input yet.'
+                'Note that lobpcg.forward does though.'
+            )
+        if A.dtype in (torch.complex64, torch.complex128) or \
+           B is not None and B.dtype in (torch.complex64, torch.complex128):
+            raise ValueError(
+                'lobpcg.backward does not support complex input yet.'
+                'Note that lobpcg.forward does though.'
+            )
+        if B is not None:
```

Comment:
I'm confused about all the tests above.
If you just move this first, then you're sure that B is None and no need to test for its type.

Reply:
It is, but it will at some point. Shall I remove it anyway?

Reply:
Depends when at some point is haha.
It does make this code a bit more complex and we cannot really test it because it's supposed to fail.

But if the new version is in the pipe and will be here soon, we can leave it as that sure.

Reply:
Yeah, it is high on my priority list, wanna push the paper.